core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMapImpl.kt

```diff
@@ -33,7 +33,46 @@ data class DistanceRangeMapImpl<T>(
      * structure (RangeMaps lib).
      */
     override fun putMany(entries: List<DistanceRangeMap.RangeMapEntry<T>>) {
+        // Unfortunately, built-in groupBy doesn't consider the original order, here we want
+        // consecutive values.
+        fun <U, V> groupByFirst(pairs: List<Pair<U, V>>): List<Pair<U, List<V>>> {
+            val result = mutableListOf<Pair<U, List<V>>>()
+            if (pairs.isEmpty()) return result
+            var nextKey = pairs[0].first
+            var nextValues = mutableListOf(pairs[0].second)
+            for ((key, value) in pairs.slice(1 ..< pairs.size)) {
+                if (key == nextKey) {
+                    nextValues.add(value)
+                } else {
+                    result.add(Pair(nextKey, nextValues))
+                    nextKey = key
+                    nextValues = mutableListOf(value)
+                }
+            }
+            result.add(Pair(nextKey, nextValues))
+            return result
+        }
+
+        fun <U, V> groupBySecond(pairs: List<Pair<U, V>>): List<Pair<V, List<U>>> {
+            val result = mutableListOf<Pair<V, List<U>>>()
+            if (pairs.isEmpty()) return result
+            var nextKey = pairs[0].second
+            var nextValues = mutableListOf(pairs[0].first)
```

Comment:
minor: Can't you mutualize the code by reversing the pairs first of providing a specific getKey/getValue ?
Not sure about the tradeoff on perf and readability.